### PR TITLE
plot g-score and learning curve

### PR DESCRIPTION
**Pull request by zhouzypaul**

### CODE REVIEW

File: actgen/plot.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	"""
+	with open(fname, 'r') as f:
+		reader = csv.reader(f)
+		next(reader, None)  # skip over the header of csv
+		reader_results = [[float(i) for i in row] for row in list(reader)]
+		columns = list(zip(*reader_results))
+	return np.array(columns)
+
+
+def prepro_g_score(directory):
```
Comment (camall3n) [inline actgen/plot.py]: Is prepro short for preprocess?

**Review by camall3n — COMMENTED** [review]
For future reference, I usually find it much easier to just use pandas and seaborn for stuff like this, rather than rolling my own loading and plotting code. Pandas has a `read_csv()` method that produces a `DataFrame` object, which you can send directly into seaborn's `relplot` function, and then seaborn can even compute confidence intervals for you automatically.

```python
import pandas as pd
import seaborn as sns

def read_csv(fname, agent):
    data = pd.read_csv(fname)
    data['agent'] = agent
    return data

data = pd.concat([read_csv(fname, agent) for fname, agent in zip(fnames, agents)])

sns.relplot(
    data=data, kind="line",
    x='episode', y='reward',
    hue='agent', style='agent'
)
```

**Comment by zhouzypaul** [comment]
Wow, thanks for the resource! That's so much easier and cleaner. 

I should really brush up on my pandas. Funny I was thinking of writing a small library of common plotting methods for my own future use. But of course, these things already exist. I guess the authors of seaborn thought about the same thing lol.

File: actgen/plot.py (hunk 1 above)
Comment (zhouzypaul) [reply actgen/plot.py]: yes, would you like to rename them to `preprocess` for clarity?

**Comment by zhouzypaul** [comment]
and could you confirm that [this](https://github.com/camall3n/actgen/pull/21/commits/933c5d1d8f2f6d45b392ca9ed93848a1bb0311d5) is what I should be doing to `epsilon` during training?

File: actgen/plot.py (hunk 1 above)
Comment (camall3n) [reply actgen/plot.py]: Yeah, I think that's better

**Comment by camall3n** [comment]
Yep, you're just setting epsilon to zero during evaluations. Looks good.

File: actgen/wrappers/duplicate_actions.py [hunk 2]
```diff
@@ -28,6 +28,14 @@ def step(self, action):
...
         original_a = math.floor(action / self.n_dup)
         return self.env.step(original_a)
+    
+    def get_duplicate_actions(self, a):
+        """
+        return all the actions that's a duplicate 'a' in a list
+        """
+        original_env_a = math.floor(a / self.n_dup)
+        all_duplicate_actions = list(range(original_env_a * self.n_dup, original_env_a * self.n_dup + self.n_dup))
+        return all_duplicate_actions
```
Comment (camall3n) [inline actgen/wrappers/duplicate_actions.py:38]: I like the env being the one to know which actions are duplicates.

File: actgen/agents/dqn.py [hunk 3]
```diff
@@ -8,18 +8,19 @@
 
 
 class DQNAgent():
-    def __init__(self, observation_space, action_space, params):
-        self.observation_space = observation_space
-        self.action_space = action_space
+    def __init__(self, env, params):
+        self.env = env
```
Comment (camall3n) [inline actgen/agents/dqn.py]: I usually try to keep the env and the agent fully separated. Is there a reason the agent needs to have access to the full env?

If this change was to compensate for the fact that now the env knows which actions are duplicates, maybe we can simply pass the `get_duplicate_actions()` function as a callable argument alongside the `observation_space` and `action_space`?

Comment (zhouzypaul) [reply actgen/agents/dqn.py]: yea that's exactly the reason. Good point! I'll keep the env and agent fully separated.

**Comment by zhouzypaul** [comment]
@camall3n is there a way to set up CUDA in GitHub actions to add unit testing for GPU computations?

**Comment by camall3n** [comment]
That's a great idea, but unfortunately it's not something github supports natively. If we wanted to provide our own cloud compute infrastructure, we could hack it with Jenkins -- apparently [that's what PyTorch and CuPy use](https://github.com/dipy/dipy/issues/2184).

The best way to ensure gpu code is correct is by having a clear plan for when we send stuff back and forth to the gpu, testing on a machine with a dedicated gpu, and then ideally not touching the code that deals with devices again.

File: actgen/agents/action_dqn.py [hunk 4]
```diff
@@ -5,37 +5,37 @@
 
 
 class ActionDQNAgent(DQNAgent):
-    def __init__(self, env, params):
-        super().__init__(env, params)
+    def __init__(self, observation_space, action_space, get_duplicate_actions, params):
```
Comment (camall3n) [inline actgen/agents/action_dqn.py]: Can we change `get_duplicate_actions` to `get_duplicate_actions_fn`?

File: actgen/agents/action_dqn.py [hunk 5]
```diff
@@ -5,37 +5,37 @@
...
+         super().__init__(observation_space, action_space, get_duplicate_actions, params)
 
     def _get_q_targets(self, batch):
         with torch.no_grad():
             # Compute Double-Q targets
-            next_states = torch.stack(batch.next_state, dim=0).float()
+            next_states = torch.stack(batch.next_state, dim=0).float().to(self.params['device'])
             aps = []
```
Comment (camall3n) [inline actgen/agents/action_dqn.py]: It's not clear what `aps` is. Can you be more descriptive with the name?

File: actgen/agents/action_dqn.py [hunk 6]
```diff
@@ -5,37 +5,37 @@
...
             for s in next_states:
                 q_values = self._get_q_values_for_state(s).squeeze().float()
                 ap = torch.argmax(q_values, dim=0)
                 aps.append(ap)
-            aps = torch.as_tensor(aps)
+            aps = torch.as_tensor(aps).to(self.params['device'])
```
Comment (camall3n) [inline actgen/agents/action_dqn.py]: I'm surprised we need a `to(device)` here. Should this be a `stack` or `concat` instead of an `as_tensor`? Each `ap` should have the same device as `q_values`, which should inherit whatever device `s` and the model use.

File: actgen/agents/dqn.py [hunk 7]
```diff
@@ -111,19 +111,20 @@ def _get_q_targets(self, batch):
...
-        return self.q(torch.as_tensor(x).float())
+        return self.q(torch.as_tensor(x).float().to(self.params['device']))
 
     def _make_qnet(self, n_features, n_actions, params):
         dropout = params['dropout_rate']
         return MLP(n_inputs=n_features,
                    n_outputs=n_actions,
                    n_hidden_layers=params['n_hidden_layers'],
                    n_units_per_layer=params['n_units_per_layer'],
+                   device=params['device'],
```
Comment (camall3n) [inline actgen/agents/dqn.py]: Rather than having `device` be an input, it's usually simpler to do:
```python
return MLP([args...]).to(device)
```

File: actgen/change_q.py [hunk 8]
```diff
@@ -53,6 +53,8 @@ def parse_args(self):
                             help='Path to the result directory to save model files')
         parser.add_argument('--tag', type=str, default='default_exp',
                             help='A tag for the current experiment, used as a subdirectory name for saving models')
+        parser.add_argument('--gpu', default=False, action='store_true',
```
Comment (camall3n) [inline actgen/change_q.py]: Let's change this arg to `--disable-gpu`, and have the default behavior automatically detect whether there's a gpu and use it if it's available, _unless_ `--disable-gpu` has been specified.

File: actgen/change_q.py [hunk 9]
```diff
@@ -127,6 +138,7 @@ def run(self):
                              n_outputs=n_outputs,
                              n_hidden_layers=self.params['n_hidden_layers'],
                              n_units_per_layer=self.params['n_units_per_layer'],
+                             device=self.params['device'],
```
Comment (camall3n) [inline actgen/change_q.py:135]: See comment about `MLP([...]).to(device)`

File: actgen/evaluate_dqn.py [hunk 10]
```diff
@@ -52,6 +52,8 @@ def parse_args(self):
                             help='Path to the result directory to save model files')
         parser.add_argument('--tag', type=str, default='default_exp',
                             help='A tag for the current experiment, used as a subdirectory name for saving models')
+        parser.add_argument('--gpu', default=False, action='store_true',
+                            help='Run training on GPU')
```
Comment (camall3n) [inline actgen/evaluate_dqn.py]: See comment about `--disable-gpu`

File: actgen/plot.py [hunk 11]
```diff
@@ -80,22 +80,28 @@ def preprocess_training_rewards(directory):
 	return time_step, avg_rewards, ci
 
 
-def plot_training_data(normal, oracle, no_duplicate, exp_name, data_type):
+def plot_training_data(normal, oracle, no_duplicate, rand, rand_oracle, exp_name, data_type):
```
Comment (camall3n) [inline actgen/plot.py:117]: This code is getting a little difficult to read. Is there an easy way to clean it up?

File: actgen/train.py [hunk 12]
```diff
@@ -104,13 +106,22 @@ def setup(self):
...
             raise RuntimeError('dqn_train_pin_other_q_values must be set to true when performing oracle action generalization')
+        
+        if self.params['gpu']:
+            if torch.cuda.is_available():
+                torch.backends.cudnn.benchmark = True
+                self.params['device'] = torch.device('cuda')
+            else:
+                raise RuntimeError('there is no GPU available')
+        else:
+            self.params['device'] = torch.device('cpu')
```
Comment (camall3n) [inline actgen/train.py]: I'm seeing the same code in at least 3 places. While you're changing it to `--disable-gpu`, can we also refactor and throw this block in utils?

File: actgen/agents/directed_qnet.py [hunk 13]
```diff
@@ -48,15 +50,15 @@ def directed_update(self, states, actions, delta_update, n_updates, baseline_qne
...
+                        original_q_values = self.forward(s.float().to(self.device))  # (1, n_actions)
+                        q_target = original_q_values.clone().to(self.device)  # (1, n_actions)
                         q_target[0][a] = float(original_q_values[0][a]) + delta_update  # (1, n_actions)
 
                         if self.pin_other_q_values:
                             get_current_q_vals = lambda s: self.forward(s.float())  # (1, n_actions)
                         else:
                             a = torch.as_tensor([a])  #(1)
-                            q_target = extract(q_target, a, idx_dim=-1)  # (1)
+                            q_target = extract(q_target, a, idx_dim=-1).to(self.device)  # (1)
```
Comment (camall3n) [inline actgen/agents/directed_qnet.py]: I wouldn't expect you to need an extra `to(device)` if all you're doing is operations on tensors that are already on the device you want.

File: actgen/agents/action_dqn.py (hunk 5 above)
Comment (zhouzypaul) [reply actgen/agents/action_dqn.py]: it's intended to be `a_prime`s, I will change it to `next_actions`

File: actgen/plot.py (hunk 11 above)
Comment (zhouzypaul) [reply actgen/plot.py:117]: yea I agree it's getting kind of confusing. I will refactor all of `plot.py` to use pandas and seaborn soon

File: actgen/change_q.py (hunk 9 above)
Comment (zhouzypaul) [reply actgen/change_q.py:135]: I think this one still might need to know the device because some tensors in `directed_qnet.py` still needs to know which device to be put on.

File: actgen/agents/directed_qnet.py [hunk 14]
```diff
@@ -11,14 +11,15 @@ class DirectedQNet(MLP):
...
-        super().__init__(n_inputs, n_outputs, n_hidden_layers, n_units_per_layer, device)
+        super().__init__(n_inputs, n_outputs, n_hidden_layers, n_units_per_layer)
         assert optim in ['sgd', 'adam']
         self.optim = optim
         self.lr = lr
         assert agent_type in ['dqn', 'action_dqn']
         self.agent_type = agent_type
         self.pin_other_q_values = pin_other_q_values
         self.device = device
+        self.model = self.model.to(device)
```
Comment (camall3n) [inline actgen/agents/directed_qnet.py]: Is this line necessary? We should be able to call `to(device)` from the outside.

File: actgen/agents/directed_qnet.py [hunk 15]
```diff
@@ -101,7 +102,7 @@ def get_current_q_vals(s):
                     new_q_values = self.forward(qnet_input.to(self.params['device'])).squeeze()  # (n_actions)
                     assert new_q_values.shape == (len(actions), )
                 # update q_delta matrix
-                q_delta[s_idx, a_idx, :] = new_q_values.detach().numpy().squeeze() - original_q_values.detach().numpy().squeeze()
+                q_delta[s_idx, a_idx, :] = new_q_values.detach().to(torch.device('cpu')).numpy().squeeze() - original_q_values.detach().to(torch.device('cpu')).numpy().squeeze()
```
Comment (camall3n) [inline actgen/agents/directed_qnet.py]: I'd prefer `.cpu()` here over `.to(torch.device('cpu'))`. I'm also not sure if it's necessary if you're calling `numpy()` right afterwards

File: actgen/change_q.py [hunk 16]
```diff
@@ -122,7 +116,7 @@ def run(self):
         s, done = self.test_env.reset(), False
         for _ in tqdm(range(self.params['n_gscore_states'])):
             # figure out what the nest state is
-            action_taken = self.agent.act(s)
+            action_taken = self.agent.act(s).to(torch.device('cpu'))
```
Comment (camall3n) [inline actgen/change_q.py]: `.cpu()`

File: actgen/evaluate_dqn.py [hunk 17]
```diff
@@ -52,8 +52,8 @@ def parse_args(self):
                             help='Path to the result directory to save model files')
         parser.add_argument('--tag', type=str, default='default_exp',
                             help='A tag for the current experiment, used as a subdirectory name for saving models')
-        parser.add_argument('--gpu', default=False, action='store_true',
-                            help='Run training on GPU')
+        parser.add_argument('--disable_gpu', default=False, action='store_true',
+                            help='enforce training on CPU')
```
Comment (camall3n) [inline actgen/evaluate_dqn.py:56]: If most of the argparse stuff is shared among all the scripts, maybe we should abstract that out at some point to make it easier to keep things up to date?

File: actgen/evaluate_dqn.py [hunk 18]
```diff
@@ -123,7 +117,7 @@ def evaluate(self):
                 for _ in range(self.params['n_eval_episodes']):
                     s, G, done, t = self.test_env.reset(), 0, False, 0
                     while not done:
-                        a = self.agent.act(s, testing=True)
+                        a = self.agent.act(s, testing=True).to(torch.device('cpu'))
```
Comment (camall3n) [inline actgen/evaluate_dqn.py]: `.cpu()`

File: actgen/train.py [hunk 19]
```diff
@@ -246,6 +248,7 @@ def run(self):
             else:
                 s = sp
             utils.every_n_times(self.params['eval_every_n_steps'], step, self.evaluate, step)
+            utils.every_n_times(self.params['eval_every_n_steps'], step, self.save_batch_loss, step, loss)
```
Comment (camall3n) [inline actgen/train.py]: If we only eval every N steps, then we only see the loss every N steps, which will be pretty noisy. Can we make a new hyperparameter `save_loss_every_n_steps = 1`?

File: actgen/agents/directed_qnet.py (hunk 15 above)
Comment (zhouzypaul) [reply actgen/agents/directed_qnet.py]: When I tried to run it, I think torch doesn't let me convert to `.numpy()` when the tensor is on the GPU.

File: actgen/agents/directed_qnet.py (hunk 14 above)
Comment (zhouzypaul) [reply actgen/agents/directed_qnet.py]: ah ok, good point. 

After I put the model on the GPU using `to(device)` from the outside, I still should put all the necessary tensors in this `DirectedQnet` class on the GPU right?

Comment (camall3n) [reply actgen/agents/directed_qnet.py]: I'm not sure I understand the question, but I think this may answer it...

Suppose you have a model called `d = DirectedQnet([args]).to(device)`, and suppose `DirectedQnet.forward()` takes input argument `x`. Since `d` is on `device`, `x` needs to be on the same `device`. If `y = d(x)`, then `y` is also on the same `device` until you call `cpu()`.

Does that help?

**Review by camall3n — APPROVED** [review]

File: actgen/agents/directed_qnet.py (hunk 14 above)
Comment (zhouzypaul) [reply actgen/agents/directed_qnet.py]: yea thanks!
